src/connection.c

```diff
@@ -154,12 +155,41 @@ mpd_connection_new(const char *host, unsigned port, unsigned timeout_ms)
 						&connection->timeout,
 						&connection->error);
 		}
+	}
 #endif
 
-		if (fd == MPD_INVALID_SOCKET)
+	if (fd == MPD_INVALID_SOCKET) {
+		/* special case: try socket in XDG_RUNTIME_DIR */
+		mpd_settings_free(settings);
+		const char *xdg_runtime_dir = getenv("XDG_RUNTIME_DIR");
+		if (xdg_runtime_dir == NULL) {
+			return connection;
```

Comment:
I'm not happy with this control flow. This may work now, but imagine we add another special case later, then it will be skipped just because there is no XDG variable, and it's hard to see from the code why. This coding style is what may cause some pain eventually, even though it may not be actually incorrect now.

Reply:
This is something I tried to address in my somewhat disastrous attempt. I think I have a better solution.

What if `mpd_connection_new()` had a list of settings structs to try instead, and move all the 'things to try' code to `mpd_settings_new()`?

E.g. add a next field like this:
```c
struct mpd_settings {
    ...
    mpd settings *next;
}
```

Then have `mpd_settings_new()` conditionally populate it, something like this:

```c
settings->host = mpd_get_xdg_socket_location();
if (settings->host != NULL) {
    settings->next = mpd_settings_new(DEFAULT SOCKET, 0, 0, NULL, NULL);
    settings->next->next = mpd_settings_new(DEFAULT_HOST, 0, 0, NULL, NULL);
}
```

Then `mpd_connection_new()` can just loop through them, like this:
```c
while (fd == MPD_INVALID_FD && (settings = mpd_settings_get_next_alternative(settings)) != NULL) {
    mpd_error_clear(&connection->error);
    fd = mpd_socket_connect(mpd_settings_get_host(settings), mpd_settings_get_port(settings), ...);
}

if (fd == MPD_INVALID_FD)
    return connection;
```

(OOM code and memory management changes omitted for brevity.)

This makes `mpd_connection_new` simpler and allows better control over the order things are attempted in. The cost is a more complex `mpd_settings_new()`.